dm/pkg/shardddl/optimism/lock.go

```diff
@@ -258,181 +247,34 @@ func (l *Lock) TrySync(info Info, tts []TargetTable) (newDDLs []string, cols []s
-		var cmp int
-		if cmp, err = nextTable.Compare(oldJoined); err == nil && cmp == 0 {
-			if col, err2 := GetColumnName(l.ID, ddls[idx], ast.AlterTableAddColumns); err2 != nil {
-				return newDDLs, cols, err2
-			} else if len(col) > 0 && l.IsDroppedColumn(info.Source, info.UpSchema, info.UpTable, col) {
-				return newDDLs, cols, terror.ErrShardDDLOptimismTrySyncFail.Generate(
-					l.ID, fmt.Sprintf("add column %s that wasn't fully dropped in downstream. ddl: %s", col, ddls[idx]))
-			}
-			newDDLs = append(newDDLs, ddls[idx])
-			continue
-		}
-
-		// try to join tables.
-		newJoined, err = joinTable(nextTable)
-		if err != nil {
-			return emptyDDLs, emptyCols, err
-		}
-
-		cmp, err = oldJoined.Compare(newJoined)
-		// FIXME: Compute DDLs through schema diff instead of propagating DDLs directly.
-		// and now we MUST ensure different sources execute same DDLs to the downstream multiple times is safe.
-		if err != nil {
-			// resolving conflict in non-intrusive mode.
-			log.L().Warn("resolving conflict", zap.String("lock", l.ID), zap.String("source", callerSource), zap.String("schema", callerSchema), zap.String("table", callerTable),
-				zap.Stringer("joined-from", oldJoined), zap.Stringer("joined-to", newJoined), zap.Strings("ddls", ddls))
-			return ddls, cols, nil
-		}
-		if cmp != 0 {
-			// < 0: the joined schema become larger after applied these DDLs.
-			//      this often happens when executing `ADD COLUMN` for the FIRST table.
-			// > 0: the joined schema become smaller after applied these DDLs.
-			//      this often happens when executing `DROP COLUMN` for the LAST table.
-			// for these two cases, we should execute the DDLs to the downstream to update the schema.
-			log.L().Info("joined table info changed", zap.String("lock", l.ID), zap.Int("cmp", cmp), zap.Stringer("from", oldJoined), zap.Stringer("to", newJoined),
-				zap.String("source", callerSource), zap.String("schema", callerSchema), zap.String("table", callerTable), zap.Strings("ddls", ddls))
-			if cmp < 0 {
-				// check for add column with a larger field len
-				if col, err2 := AddDifferentFieldLenColumns(l.ID, ddls[idx], oldJoined, newJoined); err2 != nil {
-					return ddls, cols, err2
-				} else if len(col) > 0 && l.IsDroppedColumn(info.Source, info.UpSchema, info.UpTable, col) {
-					return ddls, cols, terror.ErrShardDDLOptimismTrySyncFail.Generate(
-						l.ID, fmt.Sprintf("add column %s that wasn't fully dropped in downstream. ddl: %s", col, ddls[idx]))
-				}
-			} else {
-				if col, err2 := GetColumnName(l.ID, ddls[idx], ast.AlterTableDropColumn); err2 != nil {
-					return ddls, cols, err2
-				} else if len(col) > 0 {
-					err = l.AddDroppedColumn(info, col)
-					if err != nil {
-						log.L().Error("fail to add dropped column info in etcd", zap.Error(err))
-						return ddls, cols, terror.ErrShardDDLOptimismTrySyncFail.Generate(l.ID, "fail to add dropped column info in etcd")
-					}
-					cols = append(cols, col)
-				}
-			}
```

Comment:
these logic move to func checkAddDropColumns

Reply:
this func name is `checkAddDropColumn` 😬